include/libtorrent/packet_pool.hpp

```diff
@@ -78,78 +78,82 @@ namespace libtorrent
+	struct TORRENT_EXTRA_EXPORT packet_pool : private single_threaded
 	{
 		packet *acquire(std::size_t allocate)
 		{
 			TORRENT_ASSERT(is_single_thread());
 			TORRENT_ASSERT(allocate <= std::numeric_limits<std::uint16_t>::max());
 
-			packet *p;
+			packet_ptr p;
 
-			if (allocate <= m_syn_slab.allocate_size) { p = m_syn_slab.try_alloc(); }
-			else if (allocate <= m_mtu_floor_slab.allocate_size) { p = m_mtu_floor_slab.try_alloc(); }
-			else if (allocate <= m_mtu_ceiling_slab.allocate_size) { p = m_mtu_ceiling_slab.try_alloc(); }
-			else { p = static_cast<packet*>(malloc(sizeof(packet) + allocate)); }
+			if (allocate <= m_syn_slab.allocate_size) { m_syn_slab.try_alloc(p); }
+			else if (allocate <= m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_alloc(p); }
+			else if (allocate <= m_mtu_ceiling_slab.allocate_size) { m_mtu_ceiling_slab.try_alloc(p); }
+			else { m_rest_slab.try_alloc(p); }
 
 			p->allocated = static_cast<std::uint16_t>(allocate);
-			return p;
+			return p.release();
 		}
 
 		void release(packet *p)
 		{
 			TORRENT_ASSERT(is_single_thread());
 
 			if (p == nullptr) return;
-			std::size_t allocated = p->allocated;
 
-			if (allocated <= m_syn_slab.allocate_size) { m_syn_slab.try_free(p); }
-			else if (allocated <= m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_free(p); }
-			else if (allocated <= m_mtu_ceiling_slab.allocate_size) { m_mtu_ceiling_slab.try_free(p); }
-			else { free(p); }
+			packet_ptr pp(p); // take ownership and auto free
+			std::size_t allocated = pp->allocated;
+
+			if (allocated <= m_syn_slab.allocate_size) { m_syn_slab.try_free(pp); }
+			else if (allocated <= m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_free(pp); }
+			else if (allocated <= m_mtu_ceiling_slab.allocate_size) { m_mtu_ceiling_slab.try_free(pp); }
 		}
 	private:
-
-		template<std::size_t PACKET_LIMIT, std::size_t ALLOCATE_SIZE>
-		struct packet_slab
-		{
-			aux::vector<packet*, std::size_t> m_storage;
-			std::size_t idx{ 0 };
-			static const std::size_t allocate_size{ ALLOCATE_SIZE };
-
-			packet_slab() : m_storage(PACKET_LIMIT, nullptr) { }
-
-			~packet_slab() { flush(); }
-
-			void flush()
-			{
-				while (idx != 0) {
-					free(m_storage[--idx]);
-				}
-			}
-
-			void try_free(packet *p)
-			{
-				if (idx >= PACKET_LIMIT)
-					free(p);
-				else
-					m_storage[idx++] = p;
-			}
-
-			packet *try_alloc()
-			{
-				if (idx == 0)
-					return static_cast<packet*>(malloc(sizeof(packet) + allocate_size));
-				else
-					return m_storage[--idx];
-			}
-		};
-
-		packet_slab<0x400, sizeof(utp_header)> m_syn_slab;
-		packet_slab<0x200, (TORRENT_INET_MIN_MTU - TORRENT_IPV4_HEADER - TORRENT_UDP_HEADER)> m_mtu_floor_slab;
-		packet_slab<0x200, (TORRENT_ETHERNET_MTU - TORRENT_IPV4_HEADER - TORRENT_UDP_HEADER)> m_mtu_ceiling_slab;
+		packet_slab<sizeof(utp_header)> m_syn_slab{ 0x400 };
+		packet_slab<(TORRENT_INET_MIN_MTU - TORRENT_IPV4_HEADER - TORRENT_UDP_HEADER)> m_mtu_floor_slab{ 0x200 };
```

Comment:
is this a common packet size btw? I wouldn't expect so. Did you collect any data on packet sizes, a histogram perhaps?
Also, these limits feel pretty high to me still. on a small system this could be way too much heap to keep around.
I also think we need some mechanism to gradually shrink these caches. perhaps some periodic tick in ``utp_socket_manager`` that frees some fraction of the cached packets. That could be a separate patch though, we should just keep it in mind